File: javascript/rational-numbers/rational-numbers.js
const euclidean_gcd = (a, b) => {
	a = Math.abs(a)
	b = Math.abs(b)
	while(b) {
		[a, b] = [b, a%b]
	}
	return a
}

export class Rational {
  constructor(n,d) {
    this.numerator =d<0 ? -n :n;
    this.denominator = d<0 ? -d :d;
  }

  add(num) {
    return new Rational(num.numerator* this.denominator + num.denominator*this.numerator,  this.denominator * num.denominator).reduce();
  }

  sub(num) {
    return new Rational(num.denominator*this.numerator - num.numerator* this.denominator ,  this.denominator * num.denominator).reduce();
  }

  mul(num) {
    return new Rational(this.numerator*num.numerator, this.denominator * num.denominator).reduce();
  }

  div(num) {
    return new Rational(this.numerator*num.denominator, this.denominator * num.numerator).reduce();
  }

  abs() {
    return new Rational(Math.abs(this.numerator), Math.abs(this.denominator));
  }

  exprational(p) {
    return new Rational(Math.pow(this.numerator,p), Math.pow(this.denominator, p)).reduce();
  }

  expreal(num) {
    return Math.pow(Math.pow(num, 1/this.denominator), this.numerator);
  }

  reduce() {
    const gdc = euclidean_gcd(this.numerator, this.denominator);
    return new Rational(this.numerator/gdc, this.denominator/gdc);
  }
}

File: python/rational-numbers/rational_numbers.py
from __future__ import division
from fractions import gcd


class Rational(object):
    def __init__(self, numer, denom):
        greatest_com_denom = gcd(numer, denom)
        self.numer = numer // greatest_com_denom
        self.denom = denom // greatest_com_denom



    def __eq__(self, other):
        return self.numer == other.numer and self.denom == other.denom

    def __repr__(self):
        return '{}/{}'.format(self.numer, self.denom)

    def __add__(self, other):
        return Rational(self.numer * other.denom + self.denom*other.numer, self.denom* other.denom)

    def __sub__(self, other):
        return Rational(self.numer * other.denom - self.denom*other.numer, self.denom* other.denom)

    def __mul__(self, other):
        return Rational(self.numer * other.numer , self.denom* other.denom)

    def __truediv__(self, other):
        return Rational(self.numer * other.denom , self.denom* other.numer)

    def __abs__(self):
        return Rational(abs(self.numer), abs(self.denom))
    def __pow__(self, power):
        return Rational(self.numer**power, self.denom**power)
    def __rpow__(self, base):
        return base ** (self.numer  / self.denom)
